warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/facet/FacetHandler.java

```diff
@@ -0,0 +1,363 @@
+import datawave.ingest.data.config.NormalizedContentInterface;
+import datawave.ingest.data.config.ingest.IngestHelperInterface;
+import datawave.ingest.mapreduce.handler.ExtendedDataTypeHandler;
+import datawave.ingest.mapreduce.handler.shard.ShardIdFactory;
+import datawave.ingest.mapreduce.job.BulkIngestKey;
+import datawave.ingest.mapreduce.job.writer.ContextWriter;
+import datawave.ingest.metadata.RawRecordMetadata;
+import datawave.marking.MarkingFunctions;
+import datawave.util.StringUtils;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.security.ColumnVisibility;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapreduce.StatusReporter;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.apache.hadoop.mapreduce.TaskInputOutputContext;
+import org.apache.log4j.Logger;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class FacetHandler<KEYIN,KEYOUT,VALUEOUT> implements ExtendedDataTypeHandler<KEYIN,KEYOUT,VALUEOUT>, FacetedEstimator<RawRecordContainer> {
+    
+    private static final Logger log = Logger.getLogger(FacetHandler.class);
+    
+    /* Global configuration properties */
+    
+    public static final String FACET_TABLE_NAME = "facet.table.name";
+    public static final String FACET_LPRIORITY = "facet.table.loader.priority";
+    
+    public static final String FACET_METADATA_TABLE_NAME = "facet.metadata.table.name";
+    public static final String FACET_METADATA_LPRIORITY = "facet.metadata.table.loader.priority";
+    
+    public static final String FACET_HASH_TABLE_NAME = "facet.hash.table.name";
+    public static final String FACET_HASH_TABLE_LPRIORITY = "facet.hash.table.loader.priority";
+    
+    /* Per-datatype configuration properties */
+    
+    public static final String FACET_HASH_THRESHOLD = ".facet.hash.threshold";
+    
+    public static final String FACET_CATEGORY_DELIMITER = ".facet.category.delimiter";
+    public static final String FACET_FIELD_PREDICATE_CLASS = ".facet.field.predicate.class";
+    
+    public static final String FACET_CATEGORY_PREFIX_REGEX = "\\.facet\\.category\\.name\\..*";
+    
+    private static final Text PV = new Text("pv");
+    private static final String NULL = "\0";
+    private static final Value EMPTY_VALUE = new Value(new byte[] {});
+    
+    /* Global configuration fields */
+    
+    protected Text facetTableName;
+    protected Text facetMetadataTableName;
+    protected Text facetHashTableName;
+    
+    /* Per-datatype configuration fields */
+    
+    protected int facetHashThreshold;
+    protected String categoryDelimiter = "/";
```

Comment:
Just a nit, but '/' suggests a path of some sort which is not the case here.  This is more like a definition so perhaps colon, equals or some sort of bracket may be better.  Again, just a nit.

Reply:
@ivakegg Thanks, I understand where you're coming from. It's a nit, but it's worth doing the work up-front to get it right. 

In use, we currently have configuration property values like: 
`NETWORK_NAME/GENRES,EMBEDDED_CAST_PERSON_GENDER,RATING_AVERAGE` .

Thinking about other delimiters that make sense in this context, `;` may be the best bet from a semantic perspective:
`NETWORK_NAME;GENRES,EMBEDDED_CAST_PERSON_GENDER,RATING_AVERAGE`

(I considered '|', '#', and ';', although '|' might be better visually it is also semantically overloaded)

What do you think? The changes I need to make to address this are distributed across code, tests and configuration, so I want to make sure I get something we're satisfied with the first time.

_I won't get started about embedding structured data in a property value in that it forces you do to special parsing on the value itself - considering alternatives I think it's the lesser of evils in this particular case._

Reply:
Your point about embedding structured data in a value is a very good one.  I have been a huge contributor to this evil practice unfortunately but we all have to learn somehow.  If we are not going to use a separate property, then I vote for a semicolon I guess.

Reply:
Done. This is now a semicolon.